pyiron/vasp/structure.py

```diff
@@ -177,44 +178,51 @@ def atoms_from_string(string, read_velocities=False, species_list=None):
     n_atoms = 0
     for key in atoms_dict["species_dict"].keys():
         n_atoms += atoms_dict["species_dict"][key]["count"]
-    for i in range(position_index, position_index + n_atoms):
-        vec = list()
-        for j in range(3):
-            vec.append(float(string[i].split()[j]))
-        if atoms_dict["selective_dynamics"]:
-            selective_dynamics.append(["T" in string[i].split()[k] for k in range(3, 6)])
-        positions.append(vec)
+    try:
+        for i in range(position_index, position_index + n_atoms):
+            vec = list()
+            for j in range(3):
+                vec.append(float(string[i].split()[j]))
+            if atoms_dict["selective_dynamics"]:
+                selective_dynamics.append(["T" in string[i].split()[k] for k in range(3, 6)])
+            positions.append(vec)
+    except (ValueError, IndexError):
+        raise AssertionError("The number of positions given does not match the number of atoms")
     atoms_dict["positions"] = np.array(positions)
     if not atoms_dict["relative"]:
         if atoms_dict["scaling_factor"] > 0.:
             atoms_dict["positions"] *= atoms_dict["scaling_factor"]
         else:
             atoms_dict["positions"] *= (-atoms_dict["scaling_factor"]) ** (1. / 3.)
-
-    if not (len(atoms_dict["positions"]) == n_atoms):
-        raise AssertionError()
     velocities = list()
+    atoms = _dict_to_atoms(atoms_dict, species_list=species_list)
+    if atoms_dict["selective_dynamics"]:
+        selective_dynamics = np.array(selective_dynamics)
+        unique_sel_dyn, inverse, counts = np.unique(selective_dynamics, axis=0, return_counts=True,
+                                                    return_inverse=True)
+        count_index = np.argmax(counts)
+        atoms.add_tag(selective_dynamics=unique_sel_dyn.tolist()[count_index])
+        is_not_majority = np.arange(len(unique_sel_dyn), dtype=int) != count_index
+        for i, val in enumerate(unique_sel_dyn):
+            if is_not_majority[i]:
+                for key in np.argwhere(inverse == i).flatten():
+                    atoms.selective_dynamics[int(key)] = val.tolist()
     if read_velocities:
         velocity_index = position_index + n_atoms + 1
         for i in range(velocity_index, velocity_index + n_atoms):
             vec = list()
-            for j in range(3):
-                vec.append(float(string[i].split()[j]))
-            velocities.append(vec)
+            try:
+                for j in range(3):
+                    vec.append(float(string[i].split()[j]))
+                velocities.append(vec)
+            except IndexError:
```

Comment:
Again I believe it would be better to use an check before, rather than an except statement.